### PR TITLE
AssetManager

### PR DESCRIPTION
**Pull request by AshkanZirakzadeh**
Singleton loads all assets in the game and distributes them through loadAsset(path). 

Script parses project for all load calls and loads in advance

Assets must be placed in the assets folder, and paths to assets for loading should be declared assuming root in assets folder.

Example: If you want to fetch "src/assets/pictures/new/test.png", you only need to declare "pictures/new/test.png" as the path.

### CODE REVIEW

File: scripts/bot_assetLister.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+for root, dirnames, filenames in os.walk("."):
+    [dirnames.remove(d) for d in list(dirnames) if d in excludeDir]
+    for filename in filter(filterPyInit, filenames):
+        matches.append(os.path.join(root, filename))
+
+searchLine = "AssetManager.instance().loadAsset("
+assetLines = set()
+for match in matches:
+    file = open(match, "r")
+    for line in file:
```
Comment (seenbeen) [inline scripts/bot_assetLister.py]: Shouldn't use `file` as that's a python reserved keyword...

File: util/bot_assetManager.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    def load(self, listPath):
+        f = open(listPath+".botal","r")
+        for bpath in f:
+            path = os.path.abspath(os.path.normpath(self.assetPath + bpath.rstrip()))
+            if path in self.__assets:
+                continue
+            callback = DictUtil.tryFetch(self.__ext, os.path.splitext(path)[1])
+            asset = callback(path)
+            DictUtil.tryStrictInsert(self.__assets, path, asset)
```
Comment (seenbeen) [inline util/bot_assetManager.py]: Nitpick: combine line 20 and 21 and just use `callback(path)` directly

File: util/bot_assetManager.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    def loadCallbacks(self):
```
Comment (seenbeen) [inline util/bot_assetManager.py]: Needa get rid of this later, as this shouldn't be hard-coded...

File: util/bot_assetManager.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        self.loadTypeCallback(".jpg", AssetManager.loadIMG)
+        self.loadTypeCallback(".png", AssetManager.loadIMG)
+
+    def loadAsset(self, path):
+        return DictUtil.tryFetch(self.__assets, os.path.abspath(self.assetPath + os.path.normpath("./"+path)))
+
+    def loadTypeCallback(self, extension, callback):
+        DictUtil.tryStrictInsert(self.__ext, extension, callback)
+
+    @staticmethod
```
Comment (seenbeen) [inline util/bot_assetManager.py]: TODO: Move into render

**Comment by AshkanZirakzadeh** [comment]
Will add cleanup to backlog

**Pull request merged**